org.eclipse.corrosion/src/org/eclipse/corrosion/CorrosionPreferencePage.java

```diff
@@ -382,34 +385,60 @@ private void createCommandPathsPart(Composite container) {
 	private void installCommands() {
 		installButton.setText(Messages.CorrosionPreferencePage_installing);
 		installButton.setEnabled(false);
-		Job.create(Messages.CorrosionPreferencePage_installingRustupCargo, (ICoreRunnable) monitor -> {
-			try {
-				Bundle bundle = CorrosionPlugin.getDefault().getBundle();
-				URL fileURL = FileLocator.toFileURL(bundle.getEntry("scripts/rustup-init.sh")); //$NON-NLS-1$
-				File file = new File(new URI(fileURL.getProtocol(), fileURL.getPath(), null));
-				file.setExecutable(true);
-				String[] command = new String[] { "/bin/bash", "-c", file.getAbsolutePath() + " -y" }; //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
-				ProcessBuilder builder = new ProcessBuilder(command);
-				Process process = builder.start();
-				if (process.waitFor() == 0) {
-					CorrosionPreferenceInitializer initializer = new CorrosionPreferenceInitializer();
-					initializer.initializeDefaultPreferences();
-					Display.getDefault().asyncExec(() -> {
-						setInstallRequired(false);
-						performDefaults();
-						setValid(isPageValid());
+		ProgressMonitorDialog dialog = new ProgressMonitorDialog(this.getShell());
+		try {
+			dialog.run(true, true, monitor -> {
+				monitor.beginTask(Messages.CorrosionPreferencePage_installingRustupCargo, 15);
+				try {
+					Bundle bundle = CorrosionPlugin.getDefault().getBundle();
+					URL fileURL = FileLocator.toFileURL(bundle.getEntry("scripts/rustup-init.sh")); //$NON-NLS-1$
+					File file = new File(new URI(fileURL.getProtocol(), fileURL.getPath(), null));
+					file.setExecutable(true);
+					String[] command = new String[] { "/bin/bash", "-c", file.getAbsolutePath() + " -y" }; //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+					Process process = new ProcessBuilder(command).start();
+					BufferedReader reader = new BufferedReader(new InputStreamReader(process.getErrorStream()));
+					CompletableFuture.runAsync(() -> {
+						reader.lines().forEachOrdered(line -> {
+							monitor.subTask(line);
+							monitor.worked(1);
+						});
 					});
-					RustManager.setDefaultToolchain("beta"); //$NON-NLS-1$
-					return;
+					while (process.isAlive() && !monitor.isCanceled()) {
```

Comment:
Not sure a wait loop is what's best in general. I would personally have created a Job with `setUser(true)` and used Job.join() which would show the dialog. That way, the wait loop is managed by Platform and you don't have to write one.